### ISSUE TITLE
Invallid Excel process ID and HWND 

### ISSUE THREAD

**Issue by belgrade-dev** [xlwings/xlwings#2329]
#### xlwings 0.30.11, Excel 2019 (Office Standard 2019) build 1808, Python 3.11.3

I'm not sure if running xlwings (or automating Excel, for that matter) on Windows server machines is a supported scenario, but that's what I'm dealing with. 

Below is the sample code that illustrates the issue: Excel PID is a completely random value. I'm guessing it gets obtained from HWND using win32process.GetWindowThreadProcessId() call, but since the HWND is also invalid, by extension the PID is invalid too. Not only that, but the apps dictionary is empty too (again, probably because of those invalid handles).

I'm actually trying to instantiate a hidden app (visible=False), but I could live with the app being visible too, if that turns out to be the only way to somehow get a proper HWND+PID. Btw, setting visible=True actually doesn't make the app visible (in the real code), but that's a separate issue.

I understand that the app gets created indirectly by the COM runtime, so there's no direct relation between the process that runs xlwings and the Excel process, which makes it hard to find the PID or HWND. On a Windows server even more so, it seems.

So, is this a supported scenario? If it's not, are there any suggestions on how to get the PID, with or without xlwings?

Test case:

```python
import xlwings
xlApp = xlwings.App(visible=True, spec=None, add_book=False, impl=None)
print (f"PID: {xlApp.pid}")
print (f"HWND: {xlApp.hwnd}")
print ("Apps:")
print (xlwings.apps)
```
```
(venv) C:\Workspace>python testpid.py
PID: 1974002336
HWND: 327956
Apps:
Apps([])

(venv) C:\Workspace>python testpid.py
PID: -1238703776
HWND: 458974
Apps:
Apps([])
```

**Comment by fzumstein** [comment]
This is sort of a dupe of https://github.com/xlwings/xlwings/issues/1743

Does xlwings work on a Windows server? Yes, but as you pointed out, a few things will error out like pid or select/activate methods.

Something like this (excerpt from a pytest) does work though without issue:

```python
@fixture(scope="module")
def book(pytestconfig):
    # Setup
    test_filepath = 'test_' + pytestconfig.getoption("book")
    shutil.copyfile(pytestconfig.getoption("book"), test_filepath)
    app = xw.App(visible=False)
    wb = app.books.open(test_filepath)
    yield wb

    # Teardown
    wb.save()
    app.quit()
    for attempt in range(5):
        try:
            os.remove(test_filepath)
            break
        except Exception as e:
            time.sleep(1)
```

**Comment by belgrade-dev** [comment]
Thanks for the info. 

I'm not sure I got the point of your code sample, but I just tried to run my test case with `spec`, `add_book` and `impl` options removed and the issue is still there.

If, on the other hand, you just wanted to make a point that the app in your sample is usable -- it is also perfectly usable in the code that I'm running even when PID and HWND are invalid. 

Or you meant something else entirely?

In any case, I just need the PID for those rare cases when using the standard Python context managers (i.e. `with` statement for the app and the books) still doesn't remove the Excel instance and leaves it hanging around in memory. In such cases I use the PID with `psutil.Process.terminate()` to perform a brute-force cleanup. It's rare, but when it happens I really need to make sure there are no lingering instances.

**Comment by fzumstein** [comment]
yeah, I assumed that it wouldn't be working on your end at all. In this case, you can ignore my code sample.

fyi, nowadays, the cleanup is done automatically or can be called explicitly via:
https://docs.xlwings.org/en/latest/api/apps.html#xlwings.main.Apps.cleanup

However, it also relies on pid, obviously, so won't work either. Bottom line: yes, I'd be interested to learn why this is causing troubles on server, too:)

**Comment by belgrade-dev** [comment]
I wish I could come up with a reproducible test case. With so many moving parts (OS patches, Office patches, Python and xlwings upgrades), plus other processes running on the machine, it's difficult to maintain a stable environment. Using Docker is out of picture here, unfortunately.

I'll see if I can take the xlwings app instance I just created (its workbook name, essentially), get a list of Excel processes (PIDs, HWNDs) from the OS, and then somehow identify the process that matches my workbook name. Not sure if that's possible but I have no other ideas at the moment.

**Comment by belgrade-dev** [comment]
Here's one possible workaround. 

In a nutshell:
* The PID returned from xlwings is validated using `psutil.pid_exists()` method.
* If the PID is not valid, a workbook is needed for this workaround. This can be any workbook, even a temporary one.
* While the workbook is open, iterate through all Excel processes and files opened by those processes. If it's our Excel process, the workbook we opened should be one of those files. Take the process' PID.
 
The sample code runs fine on Windows but isn't strictly Windows-specific, I guess. Assuming that `psutil` methods behave the same way on other platforms, the workaround should there too.

```
import psutil
import xlwings

def get_excel_pid_from_workbook(workbook_path):
    for process in psutil.process_iter():
        if process.name().lower() == "excel.exe":
            try:
                for file in process.open_files():
                    if file.path.lower() == workbook_path.lower():
                        return process.pid
            except Exception:
                pass
    return None


xlApp = xlwings.App(visible=True)

workbook_path = r"c:\temp\MyWorkbook.xlsm"
wb = xlApp.books.open(workbook_path)

pid = None
if not psutil.pid_exists(xlApp.pid):
    pid = get_excel_pid_from_workbook(workbook_path)

print (f"Excel PID: {xlApp.pid}")
print (f"Excel HWND: {xlApp.hwnd}")
print ("Apps:")
print (xlwings.apps)
print (f"Fallback PID: {pid}" if pid else "Not found")
```
```
(venv) C:\Workspace>python testpid.py
Excel PID: 1171209840
Excel HWND: 3998234
Apps:
Apps([])
Fallback PID: 11004
```

**Comment by fzumstein** [comment]
Great, thanks for looking into this! I'd be really interested though to find out why the current implementation breaks on Windows server only: 

```python
    @property
    def pid(self):
        return win32process.GetWindowThreadProcessId(self.hwnd)[1]
```

Knowing whether the hwnd is wrong or if GetWindowThreadProcessId returns something wrong would already narrow it down...

**Comment by belgrade-dev** [comment]
In my environment, the HWND on the xlwings app instance is (also) wrong. If I pass that HWND to win32process.GetWindowThreadProcessId(), I get the same (wrong) PID as the one exposed on the xlwings app instance I created. So at least it's consistently wrong.

**Comment by fzumstein** [comment]
Most likely it's *just* the hwnd that's wrong.

**Comment by belgrade-dev** [comment]
Looks like, yes. How do you get it?

**Comment by fzumstein** [comment]
it's all in here (search for hwnd): 
https://github.com/xlwings/xlwings/blob/main/xlwings/_xlwindows.py
